### PR TITLE
Ex03A

### PR DESCRIPTION
**Pull request by giovannivolpe**

### CODE REVIEW

**Comment by BenjaminMidtvedt** [comment]
I'd explain why we first use `mode="valid"` and then `mode="full"` for the convolution operation.

For `plot_image(image.squeeze())`, I'd maybe add a clarification of why and what we squeeze (batch and channel dimension). Alternatively `plot_image(image[0, 0])`

(minor) In the ReLU, Downsampling, Upsampling, the description isn't consistent about the use of "we will" / "we'll"

"We can conbine" -> "We can combine".

I would have `nn.LogSoftmax(dim=1)` instead of `nn.LogSoftmax()` to be explicit about the dimension we softmax over. Also I think it will become deprecated to not specify it.

Minor, but I prefer having flatten after a conv+act block instead of after maxpool. It makes the flatten layer more directly replaceable with a global pooling layer (which is what I think @cmanzo uses in the malaria classification example).

File: Ch03_CNN/Ex03A/cnn.ipynb
Comment (BenjaminMidtvedt) [inline Ch03_CNN/Ex03A/cnn.ipynb:1]: I'd explain why we first use `mode="valid"` and then `mode="full"` for the convolution operation.

For `plot_image(image.squeeze())`, I'd maybe add a clarification of why and what we squeeze (batch and channel dimension). Alternatively `plot_image(image[0, 0])`

(minor) In the ReLU, Downsampling, Upsampling, the description isn't consistent about the use of "we will" / "we'll"

"We can conbine" -> "We can combine".

I would have `nn.LogSoftmax(dim=1)` instead of `nn.LogSoftmax()` to be explicit about the dimension we softmax over. Also I think it will become deprecated to not specify it.

Minor, but I prefer having flatten after a conv+act block instead of after maxpool. It makes the flatten layer more directly replaceable with a global pooling layer (which is what I think @cmanzo uses in the malaria classification example).

Comment (giovannivolpe) [reply Ch03_CNN/Ex03A/cnn.ipynb:1]: > I'd explain why we first use `mode="valid"` and then `mode="full"` for the convolution operation.

Good point, done

> For `plot_image(image.squeeze())`, I'd maybe add a clarification of why and what we squeeze (batch and channel dimension). Alternatively `plot_image(image[0, 0])`

Yes, I'll add it in the book. In any case, we have already done this in Chapter 2.

> (minor) In the ReLU, Downsampling, Upsampling, the description isn't consistent about the use of "we will" / "we'll"

Good, corrected, we should always use "we'll"

> "We can conbine" -> "We can combine".

Done

> I would have `nn.LogSoftmax(dim=1)` instead of `nn.LogSoftmax()` to be explicit about the dimension we softmax over. Also I think it will become deprecated to not specify it.

Done

> Minor, but I prefer having flatten after a conv+act block instead of after maxpool. It makes the flatten layer more directly replaceable with a global pooling layer (which is what I think @cmanzo uses in the malaria classification example).

Comment (BenjaminMidtvedt) [reply Ch03_CNN/Ex03A/cnn.ipynb:1]: Looks good

File: Ch03_CNN/Ex03A/cnn.ipynb [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    "We'll explore convolutions and implement some convolutional neural networks. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Convolutions in 1D\n",
+    "\n",
+    "We'll start by implementing a simple 1D convolution with a _rectangular kernel_ that works as a running average."
```
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:18]: I would add a plot

File: Ch03_CNN/Ex03A/cnn.ipynb [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+    "We'll now implement a convolutional layer in PyTorch. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Sample Image\n",
+    "\n",
+    "We start by creating a grayscale image, i.e., an image with a single color channel, stored in the `image` PyTorch array.\n",
```
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:134]: PyTorch array --> PyTorch tensor?

File: Ch03_CNN/Ex03A/cnn.ipynb [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    "conv.weight = nn.Parameter(filters)\n",
+    "\n",
+    "features_conv = conv(image)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can now implement the `plot_features()` function...\n",
```
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:221]: plot_channels()

File: Ch03_CNN/Ex03A/cnn.ipynb [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+    "model_upsample = nn.Sequential(conv, upsample)\n",
+    "features_upsample = model_upsample(image)\n",
+    "plot_channels(features_upsample[0].detach())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## CNN Architectures for Image Transformation\n",
```
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:325]: why don't you introduce deeplay here?

Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:325]: why don't you introduce deeplay here?

File: Ch03_CNN/Ex03A/cnn.ipynb (hunk 1 above)
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:18]: I would add a plot here

File: Ch03_CNN/Ex03A/cnn.ipynb (hunk 2 above)
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:134]: PyTorch array --> PyTorch tensor?

File: Ch03_CNN/Ex03A/cnn.ipynb (hunk 3 above)
Comment (cmanzo) [inline Ch03_CNN/Ex03A/cnn.ipynb:221]: plot_channels()

**Review by cmanzo — APPROVED** [review]

**Comment by giovannivolpe** [comment]
> Small typo fix. Overall nice and pedagogical.

which one is the typo?

**Comment by giovannivolpe** [comment]
@cmanzo 

Deeplay is already introduced before. This is just to explain the layers.

There'll be a figure in the book chapter. I want to keep the code simple at this moment.

I changed the rest
